network/validator/pubsub/authorized_sender_validator.go

```diff
@@ -0,0 +1,178 @@
+package validator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/rs/zerolog"
+
+	channels "github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/network"
+
+	"github.com/onflow/flow-go/model/flow"
+	cborcodec "github.com/onflow/flow-go/network/codec/cbor"
+	"github.com/onflow/flow-go/network/message"
+)
+
+func init() {
+	// initialize the authorized roles map the first time this package is imported.
+	initializeChannelToMsgCodesMap()
+}
+
+// channelToMsgCodes is a mapping of network channels to codes of the messages communicated on them. This will be used to check the list of authorized roles associated with the channel
+var channelToMsgCodes map[network.Channel][]uint8
```

Comment:
A couple of comments on this data structure:

1. Given the coupling of these codec types with the authorizer, how about moving the consts to a higher level namespace that's used by both the `AuthorizedSenderValidator` and codec. The implementation probably doesn't need to change, but at least then it's more clear that they are used for more than the encoding. That way the codec can change the implementation and maintain compatibility with the validator by providing a transparent mapping to the official const.

2. If the codec consts/functions are moved, this mapping should probably live along side that code.

Reply:
https://github.com/onflow/flow-go/pull/2317/commits/21addd94142ee7c23ad7689d3eeefca741c6ff03

Reply:
It's good to keep them all together. I'm suggesting we move the message codes out of the codec package to reduce the coupling between the specific cbor encoding currently used, and the authorizer. Right now we need a bunch of warnings to make sure if the codec changes the authorizer is updated. This creates a higher likelihood someone accidentally introduces a bug. If we decouple the two packages, that reduces the likelihood of that type of bug, and adds flexibility to make changes within the codec/use a different codec without having to modify the authorizer.

for example, we could define a new type `type MessageType uint8` that has methods for existing functionality (e.g. `String()`, `ToInterface()` and put it in a common namespace like `network/messages`. Then codec can use the new type/methods, and provide a method that returns the `MessageType` or a payload.

It's fine if you'd rather that go in a different PR. We could add it later

Reply:
There is a direct coupling at the moment from codec -> codes . The JSON codec and cbor encode the message type differen't ways. The also don't start at the same iota see [cbor](https://github.com/onflow/flow-go/blob/master/network/codec/cbor/envelope.go#L6) and [json](https://github.com/onflow/flow-go/blob/master/network/codec/json/envelope.go#L12) . This is why we need the warnings because we actually currently can not decouple the codes from the codec.

Reply:
It would be nice to create a separate issue to update the json codec to match the cbor codec and then decouple the codes. I can create and add it to the BFT messages epic.